File: app/src/main/java/com/aria/reply/AriaNotificationListener.java
package com.aria.reply;

import android.app.*;
import android.content.*;
import android.os.*;
import android.service.notification.*;
import android.text.TextUtils;
import android.app.RemoteInput;
import android.graphics.drawable.Icon;
import java.util.*;
import org.json.JSONObject;

public class AriaNotificationListener extends NotificationListenerService {
    public static final String WA="com.whatsapp";
    private static final long DEFAULT_DELAY=1800;
    private static AriaNotificationListener instance;
    private AriaStore store;
    private final Map<String,ArrayList<String>> burst=new HashMap<>();
    private final Map<String,Long> timers=new HashMap<>();
    private final Map<String,String> lastKeys=new HashMap<>();
    private Handler h;
    public static boolean live(){return instance!=null;}

    @Override public void onCreate(){super.onCreate();instance=this;store=new AriaStore(this);h=new Handler(Looper.getMainLooper());}
    @Override public void onDestroy(){instance=null;super.onDestroy();}

    @Override public void onNotificationPosted(StatusBarNotification sbn){
        if(!WA.equals(sbn.getPackageName())) return;
        Bundle e=sbn.getNotification().extras;
        String sender=e.getString(Notification.EXTRA_TITLE,"").trim();
        String text=e.getCharSequence(Notification.EXTRA_TEXT)==null?"":e.getCharSequence(Notification.EXTRA_TEXT).toString().trim();
        if(sender.isEmpty()||text.isEmpty())return;
        if(sender.equalsIgnoreCase("You")||sender.equalsIgnoreCase("Aria"))return;
        if(isSummary(text))return;

        String key=sbn.getKey()+"|"+text;
        String old=lastKeys.get(sender);
        if(key.equals(old))return;
        lastKeys.put(sender,key);

        String contact=Contacts.resolve(this,sender);
        store.addMessage(contact,"user",text);
        ArrayList<String> q=burst.computeIfAbsent(contact,k->new ArrayList<>());
        q.add(text);

        if(timers.containsKey(contact)) h.removeCallbacksAndMessages(contact);
        Runnable job=()->generate(contact);
        h.postAtTime(job,contact,SystemClock.uptimeMillis()+DEFAULT_DELAY);
        timers.put(contact,System.currentTimeMillis()+DEFAULT_DELAY);

        if(store.bool("hide_raw",false)){ try{cancelNotification(sbn.getKey());}catch(Exception ignored){} }
    }

    private boolean isSummary(String s){
        String x=s.toLowerCase(Locale.US);
        return x.matches(".*\\b\\d+\\s+(new\\s+)?messages?\\b.*") ||
               x.contains("notifications from") || x.equals("checking for new messages");
    }

    private void generate(String contact){
        ArrayList<String> q=burst.remove(contact); timers.remove(contact);
        if(q==null||q.isEmpty())return;
        String provider=store.get("provider","Groq");
        String key=new SecurePrefs(this).get(provider.equals("Gemini")?"gemini_key":"groq_key");
        String model=store.get(provider.equals("Gemini")?"gemini_model":"groq_model",
            provider.equals("Gemini")?"gemini-2.0-flash":"llama-3.1-8b-instant");
        String context=store.context(contact,16);
        String user=String.join("\n",q);
        String system="You are Aria, a concise WhatsApp assistant. Contact: "+contact+
            ". Use context below. Never claim to have sent something unless you actually did.\nCONTEXT:\n"+context+
            "\nIf the user asks for something ambiguous, ask a clarification question. "+
            "If you promise a future action, phrase the promise clearly.";
        AriaApi.call(provider,key,model,system,user,(ok,reply,error)->{
            if(!ok){store.put("last_error",error);return;}
            store.addMessage(contact,"assistant",reply);
            extractAction(contact,reply);
            if(store.bool("auto_reply",false)) sendToLatest(contact,reply);
        });
    }

    private void extractAction(String contact,String reply){
        String x=reply.toLowerCase(Locale.US);
        if(x.contains("i'll")||x.contains("i will")||x.contains("i’ll")||
           x.contains("i'll inform")||x.contains("bata dunga")||x.contains("inform kar dunga")||
           x.contains("bhej dunga")||x.contains("remind")){
            store.addAction(contact,"Aria follow-up",reply);
            Notifications.task(this,store);
        }
    }

    private void sendToLatest(String contact,String reply){
        // Find the latest WhatsApp notification with a RemoteInput action.
        try{
            for(StatusBarNotification s:getActiveNotifications()){
                if(!WA.equals(s.getPackageName()))continue;
                Bundle e=s.getNotification().extras;
                String t=e.getString(Notification.EXTRA_TITLE,"");
                if(!Contacts.resolve(this,t).equals(contact))continue;
                for(Notification.Action a:s.getNotification().actions){
                    if(a.getRemoteInputs()!=null&&a.getRemoteInputs().length>0){
                        Intent i=new Intent(); i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        RemoteInput.addResultsToIntent(a.getRemoteInputs(),i,new Bundle(){{
                            putCharSequence(a.getRemoteInputs()[0].getResultKey(),reply);
                        }});
                        a.actionIntent.send(this,0,i);
                        return;
                    }
                }
            }
        }catch(Exception e){store.put("last_error","Direct responder: "+e.getMessage());}
    }
}

File: aria-reply-native/app/src/main/java/com/aria/reply/AriaNotificationListener.kt
package com.aria.reply

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.RemoteInput
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

class AriaNotificationListener : NotificationListenerService() {

    private val executor: ScheduledExecutorService = Executors.newScheduledThreadPool(2)
    private val pendingReplies = ConcurrentHashMap<String, ScheduledFuture<*>>()
    private val inFlight = ConcurrentHashMap.newKeySet<String>()
    private lateinit var store: AriaStore

    override fun onCreate() {
        super.onCreate()
        store = AriaStore(this)
        ensureFollowUpChannel()
    }

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        val pkg = sbn.packageName
        val isSelf = pkg == packageName
        val synthetic = isSelf && sbn.notification.extras?.getBoolean("aria_synthetic_test", false) == true

        // ── Never process our own non-synthetic notifications ─────────────────
        if (isSelf && !synthetic) return

        val isWhatsApp = pkg == "com.whatsapp" || pkg == "com.whatsapp.w4b"
        if (!isWhatsApp && !synthetic) return

        val notification = sbn.notification ?: return
        val extras = notification.extras
        val title = extras.getString("android.title")?.trim().orEmpty()
        val text  = extras.getCharSequence("android.text")?.toString()?.trim().orEmpty()

        // ── Filter Aria's own sent messages surfaced by WhatsApp ──────────────
        if (title.equals("You", ignoreCase = true)) {
            store.logEvent("Ignored WhatsApp outgoing self message", null); return
        }

        // ── Filter WhatsApp group summary lines e.g. "3 new messages" ────────
        if (SUMMARY_PATTERN.matches(text)) {
            store.logEvent("Ignored WhatsApp summary notification: $text", null); return
        }

        if (title.isBlank() || text.isBlank()) return

        // ── Deduplicate by notification key (SBN key) ─────────────────────────
        val dedupeKey = sbn.key.ifBlank { "$pkg|$title|$text" }
        if (store.checkAndMarkSeen(dedupeKey)) {
            store.logEvent("Deduplicated notification key: $dedupeKey", null); return
        }

        // ── Contact resolution ────────────────────────────────────────────────
        val resolvedName = ContactResolver.resolve(this, title)
        if (resolvedName != title) {
            store.storeResolvedContact(title, resolvedName)
            store.logEvent("Contact resolved: $title → $resolvedName", null)
        }
        val displaySender = resolvedName

        // ── Content dedupe (catches WhatsApp reposting the same text under a ──
        // new notification key, which the key-based check above misses) ───────
        if (!synthetic && store.checkAndMarkSeenContent(displaySender, text)) {
            store.logEvent("Deduplicated repeated message from $displaySender", null); return
        }

        // ── Group filter ────────────────────────────────────────────────────
        val isGroup = notification.extras?.getBoolean(Notification.EXTRA_IS_GROUP_CONVERSATION, false) == true
        val groupBlocked = isGroup && !store.replyToGroups

        // ── Contact exception filter ────────────────────────────────────────
        val contactBlocked = !store.isAllowedByContactFilter(displaySender)

        store.lastCapture = "$displaySender: $text"
        store.addConversationMessage(displaySender, "user", text)
        store.logEvent("Captured ${if (synthetic) "synthetic" else "WhatsApp"} notification from $displaySender", null)

        if (groupBlocked) {
            store.logEvent("Skipped auto-reply — group replies are off ($displaySender)", null); return
        }
        if (contactBlocked) {
            store.logEvent("Skipped auto-reply — $displaySender is filtered by contact rules (${store.contactFilterMode})", null); return
        }

        // ── Optional WA notification suppression ──────────────────────────────
        if (isWhatsApp && store.suppressWaNotifications && !synthetic) {
            runCatching { cancelNotification(sbn.key) }
            postAriaSummary(displaySender, text)
        }

        val action = notification.actions?.firstOrNull { a ->
            a.remoteInputs?.any { it.allowFreeFormInput } == true
        }
        if (action == null) {
            store.lastTargetReady = false
            store.lastTargetDescription = "Notification captured, but no RemoteInput reply action was exposed."
            store.lastTargetPackage = pkg
            store.lastTargetAt = System.currentTimeMillis()
            if (!store.autoReply && !synthetic) return
            return
        }

        val remoteInput = action.remoteInputs?.firstOrNull { it.allowFreeFormInput } ?: return
        val now = System.currentTimeMillis()
        store.lastTargetReady = true
        store.lastTargetDescription = "$pkg • ${action.title ?: "Reply"} • ${remoteInput.resultKey}"
        store.lastTargetPackage = pkg
        store.lastTargetAt = now
        lastReplyAction = action
        lastRemoteInput = remoteInput
        store.logEvent("RemoteInput target detected", true)

        if (!store.autoReply && !synthetic) return

        // ── Burst gate ────────────────────────────────────────────────────────
        // Messages arriving during the delay window are already stored; the
        // eventual generation sees the full burst as conversation context.
        val conversationKey = displaySender
        if (pendingReplies.containsKey(conversationKey) || inFlight.contains(conversationKey)) {
            store.logEvent("Coalesced message into pending reply for $displaySender", null); return
        }

        // ── Seen-gate ────────────────────────────────────────────────────────
        // If Aria already replied to this sender and you haven't opened/cleared
        // that chat yet (WhatsApp cancels the notification when you do), hold
        // off replying again — the message is still captured above either way.
        if (!synthetic && store.isPendingAck(conversationKey)) {
            store.logEvent("Held reply — waiting for you to see the previous reply to $displaySender", null); return
        }

        val delay = if (synthetic) 0L else store.minDelay.coerceAtLeast(0).toLong()
        val future = executor.schedule({
            pendingReplies.remove(conversationKey)
            if (!store.autoReply && !synthetic) return@schedule
            if (!inFlight.add(conversationKey)) return@schedule
            try {
                val history = store.recentContext(conversationKey)
                val latest = history.lastOrNull { it.role == "user" } ?: return@schedule
                val context = history.dropLast(1)
                val result = AriaApi.generate(store, conversationKey, latest.text, context)
                if (!result.ok || result.reply.isBlank()) {
                    store.lastError = "${result.error} (HTTP ${result.code})"
                    store.logEvent("AI generation failed: HTTP ${result.code}", false)
                    return@schedule
                }
                // ── Follow-up tag extraction (strips [[FOLLOWUP: ...]] before sending) ─
                val (cleanedReply, commitment) = FollowUpDetector.extractAndStrip(result.reply)
                val reply = store.marker + cleanedReply

                // ── Deliver via RemoteInput ───────────────────────────────────
                val results = Bundle().apply { putCharSequence(remoteInput.resultKey, reply) }
                val fillIn  = Intent().apply { RemoteInput.addResultsToIntent(action.remoteInputs, this, results) }
                val sent = runCatching { action.actionIntent.send(this, 0, fillIn); true }.getOrDefault(false)
                if (sent) {
                    store.lastReply = reply
                    store.addConversationMessage(conversationKey, "assistant", reply)
                    store.lastError = ""
                    store.logEvent("Reply delivered through RemoteInput", true)
                    if (!synthetic) store.setPendingAck(conversationKey, true)

                    // ── Follow-up tracking ─────────────────────────────────────
                    if (commitment != null) {
                        val isNew = store.addFollowUp(conversationKey, commitment)
                        store.logEvent(if (isNew) "Follow-up task created: $commitment" else "Follow-up task updated (already had one pending for $conversationKey): $commitment", null)
                        FollowUpNotifier.refresh(this@AriaNotificationListener)
                    }
                } else {
                    store.lastError = "RemoteInput PendingIntent could not be sent"
                    store.logEvent("RemoteInput reply send failed", false)
                }
            } finally {
                inFlight.remove(conversationKey)
            }
        }, delay, TimeUnit.SECONDS)
        pendingReplies[conversationKey] = future
    }

    /**
     * Fires when a notification is cleared — either you opened the chat (WhatsApp
     * cancels its notification when the thread is read) or swiped it away. Either way,
     * treat it as "seen" and re-arm auto-reply for that sender (see the seen-gate above).
     */
    override fun onNotificationRemoved(sbn: StatusBarNotification) {
        val pkg = sbn.packageName
        val isWhatsApp = pkg == "com.whatsapp" || pkg == "com.whatsapp.w4b"
        if (!isWhatsApp) return
        val title = sbn.notification?.extras?.getString("android.title")?.trim().orEmpty()
        if (title.isBlank() || title.equals("You", ignoreCase = true)) return
        val displaySender = ContactResolver.resolve(this, title)
        if (store.isPendingAck(displaySender)) {
            store.setPendingAck(displaySender, false)
            store.logEvent("Marked $displaySender as seen — auto-reply re-armed", null)
        }
    }

    // ── Aria summary notification (replaces suppressed WA notification) ───────
    private fun postAriaSummary(sender: String, message: String) {
        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        ensureSummaryChannel(nm)
        val n = Notification.Builder(this, CHANNEL_SUMMARY)
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setContentTitle("Aria — $sender")
            .setContentText(message)
            .setAutoCancel(true)
            .build()
        nm.notify(("summary_$sender".hashCode()), n)
    }

    // ── Persistent follow-up notification ─────────────────────────────────────
    // Posting/refreshing now lives in FollowUpNotifier so both this service and the
    // Follow-ups screen (MainActivity) can keep the same notification in sync.

    private fun ensureFollowUpChannel() {
        FollowUpNotifier.ensureChannel(this)
        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        ensureSummaryChannel(nm)
    }

    private fun ensureSummaryChannel(nm: NotificationManager) {
        if (nm.getNotificationChannel(CHANNEL_SUMMARY) == null) {
            nm.createNotificationChannel(NotificationChannel(CHANNEL_SUMMARY, "Aria Message Summary", NotificationManager.IMPORTANCE_DEFAULT))
        }
    }

    companion object {
        private const val CHANNEL_SUMMARY  = "aria_summary"

        @Volatile private var lastReplyAction: Notification.Action? = null
        @Volatile private var lastRemoteInput: RemoteInput? = null

        // WhatsApp summary lines like "3 new messages" or "2 new messages from John"
        private val SUMMARY_PATTERN = Regex("^\\d+\\s+new messages?(?:\\s+from\\s+.+)?$", RegexOption.IGNORE_CASE)

        /**
         * "Die catcher" — forces Android to rebind the listener if the OS unbound it
         * (common on aggressive-battery-management OEM ROMs, which can silently drop
         * the listener so notifications stop being captured even though the toggle
         * looks fine). Safe to call anytime; requestRebind() is a no-op if already bound.
         */
        fun rebind(context: Context) {
            runCatching {
                NotificationListenerService.requestRebind(android.content.ComponentName(context, AriaNotificationListener::class.java))
                AriaStore(context).logEvent("Requested notification listener rebind", null)
            }
        }

        fun sendDirectReply(context: Context, reply: String): Boolean {
            val action = lastReplyAction ?: return false
            val ri = lastRemoteInput ?: return false
            return runCatching {
                val results = Bundle().apply { putCharSequence(ri.resultKey, reply) }
                val fillIn  = Intent().apply { RemoteInput.addResultsToIntent(action.remoteInputs, this, results) }
                action.actionIntent.send(context, 0, fillIn)
                AriaStore(context).logEvent("Direct RemoteInput action invoked", true)
                true
            }.getOrElse {
                AriaStore(context).logEvent("Direct RemoteInput action failed: ${it.message}", false)
                false
            }
        }
    }

    override fun onDestroy() {
        pendingReplies.values.forEach { it.cancel(false) }
        pendingReplies.clear(); inFlight.clear(); executor.shutdownNow(); super.onDestroy()
    }
}
